src/server_manager/web_app/digitalocean_server.ts

```diff
@@ -38,136 +39,136 @@ const DEPRECATED_API_PORT_TAG = 'apiport';
+  DROPLET_CREATED,
+  // Userspace is running (detected by the presence of tags)
+  DROPLET_RUNNING,
+  // The server has generated its management service certificate.
+  CERTIFICATE_CREATED,
   // Server is running and has the API URL and certificate fingerprint set.
-  SUCCESS,
+  COMPLETED,
   // Server is in an error state.
   ERROR,
   // Server has been deleted.
   DELETED
 }
 
+function getCompletionFraction(state: InstallState): number {
+  // Values are based on observed installation timing.
+  // Installation typically takes 90 seconds in total.
+  switch (state) {
+    case InstallState.UNKNOWN: return 0.1;
+    case InstallState.DROPLET_CREATED: return 0.5;
+    case InstallState.DROPLET_RUNNING: return 0.55;
+    case InstallState.CERTIFICATE_CREATED: return 0.6;
+    case InstallState.COMPLETED: return 1.0;
+    default: return 0;
+  }
+}
+
 export class DigitalOceanServer extends ShadowboxServer implements server.ManagedServer {
-  private eventQueue = new EventEmitter();
-  private installState: InstallState = InstallState.UNKNOWN;
+  private onDropletActive: () => void;
+  readonly onceDropletActive = new Promise<void>(fulfill => {
+    this.onDropletActive = fulfill;
+  });
+  private installState = new ValueStream<InstallState>(InstallState.UNKNOWN);
+  private readonly startTimestamp = Date.now();
 
   constructor(
       id: string, private digitalOcean: DigitalOceanSession, private dropletInfo: DropletInfo) {
     // Consider passing a RestEndpoint object to the parent constructor,
     // to better encapsulate the management api address logic.
     super(id);
     console.info('DigitalOceanServer created');
-    this.eventQueue.once('server-active', () => console.timeEnd('activeServer'));
+    // Go to the correct initial state based on the initial dropletInfo.
+    this.updateInstallState();
+    // Start polling for state updates.
     this.pollInstallState();
   }
 
-  waitOnInstall(): Promise<void> {
-    return new Promise((fulfill, reject) => {
-      // Poll this.installState for changes.  This can poll quickly as it
-      // will not make any network requests.
-      const intervalId = setInterval(() => {
-        if (this.installState === InstallState.UNKNOWN) {
-          // installState not known, wait until next retry.
-          return;
-        }
-        // State is now known, so we can stop checking.
-        clearInterval(intervalId);
-        if (this.installState === InstallState.SUCCESS) {
-          fulfill();
-        } else if (this.installState === InstallState.ERROR) {
-          reject(new errors.ServerInstallFailedError());
-        } else if (this.installState === InstallState.DELETED) {
-          reject(new errors.DeletedServerError());
-        }
-      }, 100);
-    });
+  async *monitorInstallProgress(): AsyncGenerator<number, void> {
+    for await (const state of this.installState.watch()) {
+      yield getCompletionFraction(state);
+      if (this.isInstallTerminated()) {
```

Comment:
Instead of having this check here, I think it would be clearer if in `updateInstallState` we closed the installState if `this.isInstallTerminated()`

Reply:
OK, done